### ISSUE TITLE
Before getReceipt, check DB to see if payment has been received, if not call Stripe API as a fallback + add done by API / webhook property

### ISSUE THREAD

**Issue by wanlingt** [opengovsg/FormSG#5817]

**Comment by tshuli** [comment]
On further review, the present implementation calls stripe's API to check the payment status and retrieve the receipt, after the client is redirected to the FormSG receipt endpoint. I am not sure it makes sense to switch to using the stored webhook data:
- While we might potentially save a few ms (read from DB vs round trip to stripe API), as previously discussed the webhook could arrive late + we might also need a short amount of time to update the DB, so there is no guarantee that the information in the DB is accurate at the time of retrieval.
- While we can implement stripe's API as a fallback when DB retrieval fails, we would run the risk of providing outdated information if there are existing entries from webhook (e.g. old charge which was successful but then refunded)
- Using the API on the other hand guarantees that the receipt URL we return is always up to date, since the client would hit the receipt endpoint only after the payment is successfully recorded in stripe's system.

With this in mind, I am proposing to retain the current implementation of using stripe API, for accuracy and reliability.

Welcome further discussion
@justynoh @wanlingt @timotheeg @foochifa @LinHuiqing

**Comment by timotheeg** [comment]
Thanks ShuLi. That sounds reasonable to me 👍 . Certainly for MVP it will simplify things anyway, and if we have any cause for concern, we can revisit this at a later time. I do wonder if Stripe themselves have guidance on this, that we could refer to.

As a side note, I'm thinking  another TODO for this piece is to protect the incoming endpoint with some rate limiting or caching: we shouldn't be a potential source of uncontrolled load to stripe.

**Comment by wanlingt** [comment]
Yes now I recall! The reason we called the Stripe API is because we had to retrieve the payment intent object from Stripe, in order to retrieve the associated (successful) charge and then get the receipt. 

One other thought (can't recall if we've discussed this already) - should we also update the payment object with `payment_pending` to ~~`true`~~ `false` and `payment.status` to `successful` if the receipt returns successfully? (so another way to update the status, other than using the webhook logs)

**Comment by timotheeg** [comment]
Do you mean `payment_pending:false`? 😅 

I would argue we should. We can consider the response we get from calls we make to stripe explicitly the latest authoritative information. We did discuss it from the perspective of valid transitions and idempotency, and I reckon we should use db transactions here, to ensure that whatever we read from DB and then update later on has not been updated in-between by another process.

**Comment by wanlingt** [comment]
> Do you mean payment_pending:false? 😅

oops haha yes! I've edited my previous comment.

Will create a separate issue for the updating of the DB when we retrieve the receipt

**Comment by timotheeg** [comment]
After further thought, we're still going to check the DB first:
* Optimistically we can expect the webhook to have reached in the majority of cases (we need to have a DD metric to report on that)
* We need to maintain webhook updating the DB for cases where there is catastrophic failure on the frontend and the user never gets redirected to the payment success page
* For cases where the API is exercised, we will add the API response to the payment document log

On the complexity of computing state (shared util for API and webhook):
- we want to have unidirectional consistent movement of the status pending -> charged 
- we care for the **first successful charge** to
    -  set the payment as "charged" (not final)
    - clone the pending submission to the actual submission collection (now a **complete** submission)

TODO: finalize payment status terminology and enums with @amitogp and @staceytan1998

**Comment by timotheeg** [comment]
Collaborate with owner of #5805 to come up with payment status transition logic and APIs (don't duplicate work!)
